Ugh, so my patch above still doesn't quite work on Rawhide - for some reason the input to ```bleach.clean()``` is different on Rawhide than it is on F25:

```
====================================================== FAILURES ====================================================== 
________________________________________ TestUtils.test_markup_with_bleach_1 _________________________________________ 

self = <bodhi.tests.server.test_utils.TestUtils testMethod=test_markup_with_bleach_1>                                  
clean = <MagicMock name='clean' id='140222777514576'>      

    @mock.patch('bodhi.server.util.bleach.clean')          
    @mock.patch.object(util.bleach, '__version__', u'1.4.3')                                                           
    def test_markup_with_bleach_1(self, clean):            
        """Use mocking to ensure we correctly use the bleach 1 API."""                                                 
        text = (                                           
            '# this is a header\n'                         
            'this is some **text**\n'                      
            '<script>alert("pants")</script>'              
        )                                                  
                                                           
        util.markup(None, text)                            
                                                           
        expected_text = (                                  
            u'<div class="markdown"><h1>this is a header</h1>\n<p>this is some <strong>text'                           
            u'</strong>\n&lt;script&gt;alert("pants")&lt;/script&gt;</p></div>')                                       
        expected_tags = [                                  
            "h1", "h2", "h3", "h4", "h5", "h6", "b", "i", "strong", "em", "tt", "p", "br", "span",                     
            "div", "blockquote", "code", "hr", "pre", "ul", "ol", "li", "dd", "dt", "img", "a"]                        
        # The bleach 1 API shoudl get these attrs passed.  
        clean.assert_called_once_with(expected_text, tags=expected_tags,                                               
>                                     attributes=["src", "href", "alt", "title", "class"])                             

bodhi/tests/server/test_utils.py:394:                      
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
/usr/lib/python2.7/site-packages/mock/mock.py:948: in assert_called_once_with                                          
    return self.assert_called_with(*args, **kwargs)        
/usr/lib/python2.7/site-packages/mock/mock.py:937: in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = AssertionError('Expected call: clean(u\'<div class="markdown"><h1>this is a he...', \'hr\', \'pre\', \'ul\', \'ol\', \'li\', \'dd\', \'dt\', \'img\', \'a\'])',)
from_value = None

    def raise_from(value, from_value):
>       raise value
E       AssertionError: Expected call: clean(u'<div class="markdown"><h1>this is a header</h1>\n<p>this is some <strong>text</strong>\n&lt;script&gt;alert("pants")&lt;/script&gt;</p></div>', attributes=['src', 'href', 'alt', 'title', 'cla
ss'], tags=['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'b', 'i', 'strong', 'em', 'tt', 'p', 'br', 'span', 'div', 'blockquote', 'code', 'hr', 'pre', 'ul', 'ol', 'li', 'dd', 'dt', 'img', 'a'])
E       Actual call: clean(u'<div class="markdown"><h1>this is a header</h1>\n<p>this is some <strong>text</strong>\n<script>alert("pants")</script></p></div>', attributes=['src', 'href', 'alt', 'title', 'class'], tags=['h1', 'h2', 'h3', 
'h4', 'h5', 'h6', 'b', 'i', 'strong', 'em', 'tt', 'p', 'br', 'span', 'div', 'blockquote', 'code', 'hr', 'pre', 'ul', 'ol', 'li', 'dd', 'dt', 'img', 'a'])

/usr/lib/python2.7/site-packages/six.py:718: AssertionError
```

Since the test isn't there to assert that the input is one way or the other, I will make a new patch that just lets the text be whatever it wants to be. I'm really just trying to assert that the right attributes are used.

--- a/bodhi/server/util.py
+++ b/bodhi/server/util.py
@@ -295,6 +295,28 @@ def markup(context, text):
     Return:
         basestring: HTML representation of the markdown text
     """
+
+    # determine the major component of the bleach version installed.
+    # this is similar to the approach that Pagure uses to determine the bleach version
+    # https://pagure.io/pagure/pull-request/2269#request_diff
+    bleach_major_v = int(bleach.__version__.split('.')[0])
+
+    # the only difference in the bleach API that we use between v1 and v2 is
+    # the formatting of the attributes parameter. Bleach 1 only allowed you
+    # to specify attributes to be whitelisted for all whitelisted tags.
+    # Bleach 2 requires you to specify the list of attributes whitelisted for
+    # specific tags.
+    if bleach_major_v >= 2:
+        markdown_attrs = {
+            "img": ["src", "alt", "title"],
+            "a": ["href", "alt", "title"],
+            "div": ["class"],
+        }
+    else:
+        markdown_attrs = [
+            "src", "href", "alt", "title", "class"
+        ]
+
     markdown_tags = [
         "h1", "h2", "h3", "h4", "h5", "h6",
         "b", "i", "strong", "em", "tt",
@@ -303,10 +325,6 @@ def markup(context, text):
         "ul", "ol", "li", "dd", "dt",
         "img",
         "a",
-    ]
-
-    markdown_attrs = [
-        "src", "href", "alt", "title", "class"
     ]
 
     markdown_text = markdown.markdown(text, extensions=['markdown.extensions.fenced_code'])

--- a/bodhi/tests/server/test_utils.py
+++ b/bodhi/tests/server/test_utils.py
@@ -357,7 +357,7 @@ class TestUtils(unittest.TestCase):
         for element in result:
             assert isinstance(element, unicode)
 
-    def test_markup(self):
+    def test_markup_escapes(self):
         """Ensure we correctly parse markdown & escape HTML"""
         text = (
             '# this is a header\n'
@@ -370,6 +370,46 @@ class TestUtils(unittest.TestCase):
             '<p>this is some <strong>text</strong>\n'
             '&lt;script&gt;alert("pants")&lt;/script&gt;</p></div>'
         ), html
+
+    @mock.patch('bodhi.server.util.bleach.clean', return_value='cleaned text')
+    @mock.patch.object(util.bleach, '__version__', u'1.4.3')
+    def test_markup_with_bleach_1(self, clean):
+        """Use mocking to ensure we correctly use the bleach 1 API."""
+        text = '# this is a header\nthis is some **text**'
+
+        result = util.markup(None, text)
+
+        self.assertEqual(result, 'cleaned text')
+        expected_text = (
+            u'<div class="markdown"><h1>this is a header</h1>\n<p>this is some <strong>text'
+            u'</strong></p></div>')
+        expected_tags = [
+            "h1", "h2", "h3", "h4", "h5", "h6", "b", "i", "strong", "em", "tt", "p", "br", "span",
+            "div", "blockquote", "code", "hr", "pre", "ul", "ol", "li", "dd", "dt", "img", "a"]
+        # The bleach 1 API shoudl get these attrs passed.
+        clean.assert_called_once_with(expected_text, tags=expected_tags,
+                                      attributes=["src", "href", "alt", "title", "class"])
+
+    @mock.patch('bodhi.server.util.bleach.clean', return_value='cleaned text')
+    @mock.patch.object(util.bleach, '__version__', u'2.0')
+    def test_markup_with_bleach_2(self, clean):
+        """Use mocking to ensure we correctly use the bleach 2 API."""
+        text = '# this is a header\nthis is some **text**'
+
+        result = util.markup(None, text)
+
+        self.assertEqual(result, 'cleaned text')
+        expected_text = (
+            u'<div class="markdown"><h1>this is a header</h1>\n<p>this is some <strong>text'
+            u'</strong></p></div>')
+        expected_tags = [
+            "h1", "h2", "h3", "h4", "h5", "h6", "b", "i", "strong", "em", "tt", "p", "br", "span",
+            "div", "blockquote", "code", "hr", "pre", "ul", "ol", "li", "dd", "dt", "img", "a"]
+        expected_attributes = {
+            "img": ["src", "alt", "title"], "a": ["href", "alt", "title"], "div": ["class"]}
+        # The bleach 2 API shoudl get these attrs passed.
+        clean.assert_called_once_with(expected_text, tags=expected_tags,
+                                      attributes=expected_attributes)
 
     def test_rpm_header(self):
         h = util.get_rpm_header('libseccomp')

--- a/docs/release_notes.rst
+++ b/docs/release_notes.rst
@@ -21,6 +21,8 @@ Features
   (`#461 <https://github.com/fedora-infra/bodhi/issues/461>`_).
 * Pending releases are now displayed on the home page
   (`#1619 <https://github.com/fedora-infra/bodhi/issues/1619>`_).
+* Links without an explicit scheme can now be detected as links
+  (`#1721 <https://github.com/fedora-infra/bodhi/issues/1721>`_).
 
 
 Bugs
@@ -44,6 +46,8 @@ Bugs
   (`#1659 <https://github.com/fedora-infra/bodhi/issues/1659>`_).
 * The UI now properly urlencodes search URLs to properly escape characters such as "+"
   (`#1015 <https://github.com/fedora-infra/bodhi/issues/1015>`_).
+* e-mail addresses are now properly processed by the markdown system
+  (`#1656 <https://github.com/fedora-infra/bodhi/issues/1656>`_).
 
 
 Development improvements
@@ -67,6 +71,8 @@ Development improvements
   refactor, Bodhi's tests now consume about 450 MB instead of about 4.5 GB. As a result, the example
   Vagrantfile now uses 2 GB of RAM instead of 5 GB. It is likely possible to squeeze it down to 1 GB
   or so, if desired.
+* Bodhi now supports both the bleach 1 and bleach 2 APIs
+  (`#1718 <https://github.com/fedora-infra/bodhi/issues/1718>`_).
 
 
 Release contributors

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,6 @@
 arrow
 bunch
-bleach<2
+bleach
 click
 colander
 cornice<2

